docs/custom_promise_types/cfengine.sh

```diff
@@ -0,0 +1,164 @@
+#!/bin/false
+#
+# This file should be sourced, not run
+
+reset_state() {
+    # Set global variables before we begin another request
+
+    # Variables parsed directly from request:
+    request_operation=""
+    request_log_level=""
+    request_promise_type=""
+    request_promiser=""
+
+    for var in $(env | cut -d '=' -f 1 | grep '^request_attribute_'); do
+        unset $var
+    done
+
+    # Variables to put into response:
+    response_result=""
+
+    # Other state:
+    saw_unknown_key="no"
+    saw_unknown_attribute="no"
+
+    unknown_attribute_names=""
+}
+
+handle_input_line() {
+
+    # Split the line of input on the first '=' into 2 - key and value
+    IFS='=' read -r key value <<< "$1"
+
+    case "$key" in
+        operation)
+            request_operation="$value" ;;
+        log_level)
+            request_log_level="$value" ;;
+        promise_type)
+            request_promise_type="$value" ;;
+        promiser)
+            request_promiser="$value" ;;
+        attribute_*)
+            attribute_name=${key#"attribute_"}
+            if ! expr " $obligatory_attributes $optional_attributes " : ".* $attribute_name " >/dev/null; then
+                saw_unknown_attribute="yes"
+                unknown_attribute_names="$unknown_attribute_names, $attribute_name"
+            fi
+            eval "request_${key}=\$value" ;;
+        *)
+            saw_unknown_key="yes" ;;
+    esac
+}
+
+receive_request() {
+    # Read lines from input until empty line
+    # Call handle_input_line for each non-empty line
+    while IFS='$\n' read -r line; do
+        if [ "x$line" = "x" ] ; then
+            break
+        fi
+        handle_input_line "$line" # Parses a key=value pair
+    done
+}
+
+write_response() {
+    echo "operation=$request_operation"
+    echo "result=$response_result"
+    echo ""
+}
+
+operation_terminate() {
+    response_result="success"
+    type do_terminate >/dev/null && do_terminate
+    write_response
+    exit 0
+}
+
+operation_validate() {
+    response_result="valid"
+    if [ "$saw_unknown_attribute" != "no" -a "$all_attributes_are_valid" != "yes" ] ; then
+        echo "log_error=Unknown attribute/s: ${unknown_attribute_names#, }"
+        response_result="invalid"
+    fi
+
+    if [ ! -z "$obligatory_attributes" ]; then
+        for attribute_name in $obligatory_attributes; do
+            # Note: ${!varname} synthax expands to value of variable, which name
```

Comment:
s/synthax/syntax/ :)

Reply:
Or were you referring to the Roland AX-Synth? ;) JK.

Reply:
I make this mistake so often I should add it to OS-wide auto-correct :)